### DISCUSSION TITLE
time_pulse_us precision problem

### DISCUSSION THREAD

**Discussion by toadzhou** [micropython/micropython#10046] [Using MicroPython] [answered]
I use different GPIOs on the same ESP32 to simulate the PWM capture experiment, but when using the time_pulse_us method, his time always deviates by 0.2ms. I don't know why, I need to manually compensate for 0.2ms.

my wiring diagram
![image](https://user-images.githubusercontent.com/1249205/203244259-6f3467f2-91d9-47f6-8225-34546f7676e9.png)

```
from micropython import alloc_emergency_exception_buf
from machine import Pin,PWM,time_pulse_us
import time

alloc_emergency_exception_buf(100)

def irqtime(pin):
    a=time_pulse_us(pin,1)
    ts=round(a/1000,1)+0.2   #Here I compensated 0.2
    pwms = int(ts / 20 * 1023)
    pwm14.duty(pwms)
    print(a,ts,pwms)
    return None
   
pwm13 = PWM(Pin(13,Pin.OUT))
pwm13.freq(50)



pwm14 = PWM(Pin(14,Pin.OUT))
pwm14.freq(50)


p12 = Pin(12,Pin.IN)
p12.init(p12.IN, p12.PULL_DOWN)
p12.irq(handler=irqtime, trigger=Pin.IRQ_RISING)
for i in range(2):
    print(11)
    pwm13.duty(int((0 / 180 * 2 + 0.5) / 20 * 1023))
    time.sleep(1)
    pwm13.duty(int((90 / 180 * 2 + 0.5) / 20 * 1023))
    time.sleep(1)
    pwm13.duty(int((180 / 180 * 2 + 0.5) / 20 * 1023))
    time.sleep(1)
    print(22)
pwm13.deinit()
```

**Comment by peterhinch** [comment, marked as answer]
You start timing the pulse in the interrupt service routine (ISR) for the same pulse. When the low to high transition occurs, some time will elapse before the ISR runs. Then the call to `time_pulse_us` occurs and the method starts the timer. So my guess is that the 200μs is a measure of that latency. ESP32 only supports soft interrupts and latency can be quite long. I'm actually rather impressed that you're only seeing 200μs...

**Reply by toadzhou** [reply]
Is there any better way to handle these errors?

**Reply by peterhinch** [reply]
Calling `time_pulse_us` in an ISR is not a good idea. This is because the method blocks for the duration of the pulse (it can't produce a result until the pulse has ended). Calling blocking methods in an ISR is bad practice because it can prevent other ISR's from running. So it may be best to call it from your main code.

Loop until the pin goes low, then call the method and wait for the result. That should produce an accurate result.

**Reply by toadzhou** [reply]
Thanks, I execute it through Timer, and he can get the correct value. thank you very much!
